feat: add interruption framework for removing effects from actors

LordMidas opened the pull request:
This is my implementation of a [feature request](https://github.com/MSUTeam/MSU/issues/327) that was originally made over at MSU. If we like this to be part of MSU we could move it there.

The implementation's meat is in the `skill_container` hook where we hook the `removeByID` function and detect a vanilla "interruption" of the actor. This means that all vanilla skills which interrupt actors are automatically covered, and even modded skills which follow the vanilla style of interruption will be covered.

LordMidas committed:
feat: add interruption framework for removing effects from actors

LordMidas committed:
feat: add interruption framework for removing effects from actors

Darxo commented on mod_modular_vanilla/hooks/skills/skill_container.nut:
```diff
@@ -0,0 +1,47 @@
+				case "effects.shieldwall":
+				case "effects.spearwall":
+				case "effects.riposte":
+					local frame = ::Time.getFrame();
```
How does `getFrame()` work exactly? What does it do?
I don't have any experience with that function.
I see it being used a few times in vanilla and being implemented in the .exe.

LordMidas replied:
This is also my first time working with this function. In my testing in-game it seems to me that it prints the total number of "frames" of the game that have run since the start of the game. And I assume that a given sequence of code and function calls in squirrel runs inside a single frame.

LordMidas commented on mod_modular_vanilla/hooks/skills/effects/disarmed_effect.nut:
```diff
@@ -0,0 +1,15 @@
+			__original();
+			if (!this.isGarbage())
+			{
+				this.getContainer().getActor().interrupt(true, false);
```
The concern is that the main thing that Disarmed accomplishes is setting `properties.IsAbleToUseWeaponSkills` to false. There can be other effects which do this. We even added an `actor.isDisarmed()` function in MSU so that modders can extend that to add more effects which may accomplish something which can be considered as "disarmed" for an actor. Therefore, I do not want to assume that the `disarmed_effect` is an edge case that we can ignore in this framework. Instead I'd like a solution which covers all bases.

Your points are fair. But, as described above, I am also not fully satisfied with an interrupt framework which isn't general enough to cover these cases. 

The current implementation isn't perfect. So I'd like more suggestions on how this can be implemented in an elegant way which would achieve the goals of:
- Be general and extensible.
- Be intuitive to use.

Darxo replied:
`actor.isDisarmed()` from MSU should be sufficient. 
Any effect that is bound to the presence of a weapon (Riposte, Spearwall) should check whether this actor is currently disarmed to then remove the effect
My suggestion above of focussing just on the disarmed_effect is not nearly as flexible for this, in comparison.

Darxo commented on mod_modular_vanilla/hooks/skills/skill_container.nut:
```diff
@@ -0,0 +1,47 @@
+// MV: Added
+// Part of the actor.interrupt framework
+::MSU.Skills.addEvent("onActorInterrupted", function( _offensive, _defensive ) {});
```
I dislike how this hides the event function declaration, that you would usually expect in the `skill.nut`.
One way I mod is that I look at the base class to see what it offers. When I work on skills, I am usually only interested in what the skill.nut of vanilla or my framework mods do.

Now I also have to look in the skill_container to learn the full feature-set of skill.nut. Or where ever else that event is added.

LordMidas committed:
refactor: declare event in both skill and skill_container

Instead of using MSU addEvent function.

LordMidas committed:
refactor: declare event in both skill and skill_container

Instead of using MSU addEvent function.

LordMidas committed:
feat: prefix function names with MV_

LordMidas committed:
feat: prefix function names with MV_

LordMidas committed:
feat: add character property for immunity to interruption

LordMidas committed:
feat: add character property for immunity to interruption

LordMidas committed:
feat: remove parameters from interrupt and add MV_onDisarmed

LordMidas committed:
feat: remove parameters from interrupt and add MV_onDisarmed

LordMidas commented:
I have updated this PR with suggestions from the review and from your PR #9, along with some other changes which I think streamline the disarmed situation:
- Remove parameters from the interrupt function.
- Add character property for immunity to interruption.
- Add new `MV_onDisarmed` event for skills.
- Prefix stuff with `MV_`.

Darxo commented on mod_modular_vanilla/hooks/entity/tactical/actor.nut:
```diff
@@ -46,6 +46,14 @@
 
 ::ModularVanilla.MH.hook("scripts/entity/tactical/actor", function (q) {
+	// MV: Added
+	// Part of the actor.MV_interrupt framework
```
```suggestion
	// Part of the actor.MV_interrupt framework
	// Cause an 'Interruption' on this actor. That is a new standardized event, which will remove some temporary skills on the target. In Vanilla those interruptible skills are shieldwall, riposte and spearwall
```

I suggest we give more context here for modder, as this is the most public facing function of this framework

LordMidas committed:
fix: remove spearwall and riposte when disarmed

Co-authored-by: Darxo <Darxo@users.noreply.github.com>

LordMidas committed:
fix: remove spearwall and riposte when disarmed

Co-authored-by: Darxo <Darxo@users.noreply.github.com>

LordMidas committed:
docs: improve comment about interruption framework

Co-authored-by: Darxo <Darxo@users.noreply.github.com>

LordMidas committed:
docs: improve comment about interruption framework

Co-authored-by: Darxo <Darxo@users.noreply.github.com>

LordMidas committed:
feat(chg): rename things to include Skills in names for clarity

LordMidas committed:
feat(chg): rename things to include Skills in names for clarity

Darxo commented on mod_modular_vanilla/hooks/skills/skill_container.nut:
```diff
@@ -1,88 +1,165 @@
+
+			__original(_skillID);
+
+			if (this.m.__MV_InterruptionCount == 3)
```
This could result in false positives, if a skill removes the same effect multiple times.
For example if three compatible mods all decide, that the same skill now removes shieldwall only (but nothing else)

LordMidas committed:
feat: remove immunity to interruption character property

Because of the way we detect auto-interruption in skill_container, we cannot have support an immunity property as we have to detect removal in order to detect interruption.

LordMidas committed:
fix: store skill id instead of integer to prevent false positives
